upstairs/src/volume.rs

```diff
@@ -1092,15 +1094,355 @@ impl Volume {
+                o_opts.cert_pem,
+                n_opts.cert_pem
+            )
+        }
+        if o_opts.key_pem != n_opts.key_pem {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts key_pem invalid {:?} vs. {:?}",
+                o_opts.key_pem,
+                n_opts.key_pem
+            )
+        }
+        if o_opts.root_cert_pem != n_opts.root_cert_pem {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts root_cert_pem invalid {:?} vs. {:?}",
+                o_opts.root_cert_pem,
+                n_opts.root_cert_pem
+            )
+        }
+        if o_opts.control != n_opts.control {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts control invalid {:?} vs. {:?}",
+                o_opts.control,
+                n_opts.control
+            )
+        }
+        if o_opts.read_only != n_opts.read_only {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts read_only invalid {:?} vs. {:?}",
+                o_opts.read_only,
+                n_opts.read_only
+            )
+        }
+        if o_opts.target == n_opts.target {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts target unchanged {:?} vs. {:?}",
+                o_opts.target,
+                n_opts.target
+            )
+        }
+
+        // Walk the targets, we can only have one different
+        let mut new_target_cid = None;
+        for cid in 0..3 {
+            if o_opts.target[cid] != n_opts.target[cid] {
+                if new_target_cid.is_some() {
+                    crucible_bail!(
+                        ReplaceRequestInvalid,
+                        "sub_volume multiple different targets: {:?} vs. {:?}",
+                        o_opts.target,
+                        n_opts.target
+                    )
+                } else {
+                    new_target_cid = Some(cid);
+                }
+            }
+        }
+
+        // We should not exit the loop without setting new_target_cid
+        if let Some(cid) = new_target_cid {
+            Ok((o_opts.target[cid], n_opts.target[cid]))
+        } else {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume no target difference {:?} vs. {:?}",
+                o_opts.target,
+                n_opts.target
+            )
+        }
+    }
+
+    // Given two VolumeConstructionRequests, compare them to verify they
+    // only have the proper differences and if the VCRs are valid, submit
+    // the targets for replacement.  A success here means the upstairs has
+    // accepted the replacement and the process has started.
+    pub async fn target_replace(
+        &self,
+        original: VolumeConstructionRequest,
+        replacement: VolumeConstructionRequest,
+        log: &Logger,
```

Comment:
Is this the interface Propolis is going to call? It looks great, except that we don't have a logger handy in the `CrucibleBackend`, because the backend doesn't hang onto the logger that it passed down to the upstairs in `Volume::construct`. Can the top-level Volume hold onto that logger in addition to passing it down to the subvolumes so that the caller doesn't have to construct a new one here?

Reply:
Yeah, this interface was what I was hoping Propolis could use.

I've wanted a few times to add a `Logger` field to the volume layer, but so far have not done so.  Maybe this is the time to do that.  Let me see if I can figure out how much work it is to do that.  I think since we have updated Propolis to supply a logger at `Volume::construct`, crucible can attach it there.  It's really all the test code I'll have to change :)